### ISSUE TITLE
NextJS does not support moduleResolution: Node16

### ISSUE THREAD

**Issue by pm0u** [vercel/next.js#44095]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.0.7
      eslint-config-next: 13.0.6
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

TypeScript

### Link to the code that reproduces this issue

https://github.com/pm0u/nextjs-module-resolution

### To Reproduce

run yarn build, see that it fails
Go to the `pages/index.tsx` file in VSCode, remove the `//@ts-expect-error` and see that there's an error without it

### Describe the Bug

I found this because when set to Node16, there is an error with a library we import so I set a `//@ts-expect-error` above that import. This seemed to work fine in dev mode, so I thought it was an issue with the package maintainers. It seems the build proceeds because NextJS forces `moduleResolution: "node"` which allows the import to work.

### Expected Behavior

NextJS would use the tsconfig that I provide in it's entirety.

Related issue: https://github.com/vercel/next.js/issues/35572

We are now at TS 4.9.x so this module resolution is well out of beta as was originally discussed in that issue.

### How are you deploying your application? (if relevant)

next build

**Comment by pm0u** [comment]
Also related:

https://github.com/vercel/next.js/issues/38854
https://github.com/vercel/next.js/issues/37525
https://github.com/vercel/next.js/issues/40854
https://github.com/vercel/next.js/issues/39921

**Comment by bluelovers** [comment]
https://github.com/bluelovers/with-mongodb

`dev` is ok
`build` is failed

```
<html>
<body>
<!--StartFragment-->

info  - Linting and checking validity of types...
--
15:38:09.644 | We detected TypeScript in your project and reconfigured your tsconfig.json file for you. Strict-mode is set to false by default.
15:38:09.645 |  
15:38:09.645 | The following mandatory changes were made to your tsconfig.json:
15:38:09.645 |  
15:38:09.645 | - moduleResolution was set to node (to match webpack resolution)
15:38:09.645 |  
15:38:09.722 |  
15:38:09.723 | > Build error occurred
15:38:09.724 | Error: Debug Failure. Unexpected moduleResolution: node
15:38:09.724 | at resolveModuleName (/vercel/path0/node_modules/typescript/lib/typescript.js:36563:24)
15:38:09.724 | at Object.resolve (/vercel/path0/node_modules/typescript/lib/typescript.js:101391:47)
15:38:09.724 | at loadWithModeAwareCache (/vercel/path0/node_modules/typescript/lib/typescript.js:101431:40)
15:38:09.724 | at actualResolveModuleNamesWorker (/vercel/path0/node_modules/typescript/lib/typescript.js:101647:126)
15:38:09.724 | at resolveModuleNamesWorker (/vercel/path0/node_modules/typescript/lib/typescript.js:101924:22)
15:38:09.725 | at resolveModuleNamesReusingOldState (/vercel/path0/node_modules/typescript/lib/typescript.js:102066:77)
15:38:09.725 | at processImportedModules (/vercel/path0/node_modules/typescript/lib/typescript.js:103385:29)
15:38:09.725 | at findSourceFileWorker (/vercel/path0/node_modules/typescript/lib/typescript.js:103178:9)
15:38:09.725 | at findSourceFile (/vercel/path0/node_modules/typescript/lib/typescript.js:103054:22)
15:38:09.725 | at /vercel/path0/node_modules/typescript/lib/typescript.js:103019:24 {
15:38:09.726 | type: 'Error'
15:38:09.726 | }
15:38:09.747 | error Command failed with exit code 1.
15:38:09.748 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
15:38:09.764 | Error: Command "yarn run build" exited with 1

<!--EndFragment-->
</body>
</html>
```

**Comment by cseas** [comment]
Next.js doesn't support `moduleResolution: bundler` either. These new moduleResolution options are important for users trying to consumer pure ESM libraries.

**Comment by balazsorban44** [comment]
I checked your reproduction, and the issue here is that you are missing `"type": "module"` from your `package.json`.

The error should tell you this:

> The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("next-sanity")' call instead.
  To convert this file to an ECMAScript module, add the field `"type": "module"` to '/yourproject/package.json'.ts(1479)

Since `@ts-expect-error` makes the build pass, it means that Next.js is actually fixing the issue for you, instead of failing.

Another solution would be to use the `.mts` extension to force ESM on your page file, but since it has JSX, it should technically be `.mtsx` but there is no such file extension supported by TS.

You can verify it's not a Next.js bug, by running your `npm run types` script, which would then use `tsc`, totally skipping Next.js.

Adding `"type": "module"` to your package.json file fixes the `next-sanity` import. There might be a related (but different) issue with importing `next/*` submodules though: https://github.com/vercel/next.js/issues/46078

Since this is not a Next.js bug, I'm closing it for now. :+1:

**Comment by pm0u** [comment]
@balazsorban44 while that technically solves the issue you highlighted, it does not solve the larger issue of NextJS modifying my `tsconfig.json` at build or otherwise not respecting settings there.

> Since @ts-expect-error makes the build pass, it means that Next.js is actually fixing the issue for you, instead of failing.

I didn't include the entire list of dependencies in this example, but we had other dependencies that relied on `moduleResolution: 'node16'`. While it does "fix" the import issue (see above, I don't think this is a fix) it breaks many other things.

> You can verify it's not a Next.js bug, by running your npm run types script, which would then use tsc, totally skipping Next.js.

This is the other issue, this is not a positive thing - there is no way to replicate the type check that NextJS will use. `tsc` passes, next does not.

> Adding "type": "module" to your package.json file fixes the next-sanity import. There might be a related (but different) issue with importing next/* submodules though: https://github.com/vercel/next.js/issues/46078

I don't see how this is a fix if it breaks built in Next features.

---

tl;dr - this issue could be interpreted as one or both of two issues I suppose:

- NextJS not respecting all settings in `tsconfig.json` - which basically means any and all type checks and IDE hints are to be questioned.
- NextJS not providing a way to do a 1:1 type check in line with the build type check - so, I have no way to know that the settings in the repository would break at build, it cannot be caught via CI, cannot be caught via pre-commit, etc.

For the first point, that is covered in the linked issues so in that case closing this could make sense however this is not in fact resolved. For the second point, I do not see any resolution here or in the linked issues.
